crates/ethrpc/src/current_block/mod.rs

```diff
@@ -143,7 +145,7 @@ pub fn into_stream(receiver: CurrentBlockStream) -> WatchStream<BlockInfo> {
 /// Trait for abstracting the retrieval of the block information such as the
 /// latest block number.
 #[async_trait::async_trait]
-pub trait BlockRetrieving: Send + Sync + 'static {
+pub trait BlockRetrieving: Debug + Send + Sync + 'static {
```

Comment:
Adding `Debug` here seems like a code smell. I think we can drop the trait altogether and only have a single type we pass around instead. This refactor looks to be pretty straight forward but further increase the scope of this PR.

Reply:
We currently have two implementations of BlockRetrieving (one based on Web3 and one based on making a smart contract call to return `latest.block`). I think the latter is not working anymore since Nethermind started to return latest block information even when called with `pending` tag (thus always returning block - 1), so maybe it's safe to remove (other nodes may still support it, but we were told it would go away eventually).

I'm wondering why do you think requiring Debug here is a smell? I can also override the Debug implementation of the component that is being debug printed and just print some generic message for block retrieving if that's better.

Reply:
> We currently have two implementations of BlockRetrieving

Yeah, we use one and the one we use uses the other under the hood. So instead of making `Web3` implement `BlockRetrieving` we can simply have internal functions that fetch the block using `Web3` instead.

> I'm wondering why do you think requiring Debug here is a smell?

The functionality of `BlockRetrieving` does in no way depend on the implementing type also implementing `Debug`. The (unfortunately) idiomatic? way of doing this would be to have `BlockRetrieving` be pure (i.e. not add constraints it doesn't need for it's implementation) and have another trait that is `PrintableBlockRetriving: BlockRetrieving + Debug` where ever you need that.
This is pretty clumsy so I though the nicest way to deal with that is to ditch the trait altogether.

I think you don't have to revert the `Debug` change here since it's not a big deal and should be going away eventually anyway. Just wanted to mention that `BlockRetrieving` does not depend on `Debug` to work correctly.

Reply:
Agree with @MartinquaXD that this is a smell related to how we use trait objects. In particular, as @MartinquaXD suggested, `BlockRetrieving` does not need `Debug` to work (this is different some other situation where you really want a trait to indicate that you need the existing functionality of its super-trait, plus some other methods).

I think the "idiomatic" way is to have a separate trait that is auto-implemented for all types that implement both traits:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=38d646211165ff54c936993f7a20e490

This is, IMO, **overkill**. The codesmell is fine and a downside of using trait-objects that the codebase just has to live with given its existing design choices. This is, IMO, the same codesmell to the `Send + Sync + 'static` that we tack on all of our traits we use with trait objects to make using them easier.